packages/editor/src/extensions/link/helpers/pasteHandler.ts

```diff
@@ -20,32 +20,65 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 import { Editor } from "@tiptap/core";
 import { MarkType } from "@tiptap/pm/model";
 import { Plugin, PluginKey } from "@tiptap/pm/state";
+import linkifyHtml from "linkify-html";
 import { find } from "linkifyjs";
+import { linkRegex } from "../link";
 
 type PasteHandlerOptions = {
   editor: Editor;
   type: MarkType;
+  linkOnPaste: boolean;
 };
 
 export function pasteHandler(options: PasteHandlerOptions): Plugin {
+  let shiftKey = false;
   return new Plugin({
     key: new PluginKey("handlePasteLink"),
     props: {
+      handleKeyDown(_, event) {
+        shiftKey = event.shiftKey;
+        return false;
+      },
       handlePaste: (view, event, slice) => {
+        const clipboardHtmlData = event.clipboardData?.getData("text/html");
+        if (clipboardHtmlData) {
+          return false;
+        }
+
         const { state } = view;
         const { selection } = state;
         const { empty } = selection;
 
-        if (empty) {
-          return false;
-        }
-
         let textContent = "";
 
         slice.content.forEach((node) => {
           textContent += node.textContent;
         });
 
+        // don't deal with markdown links in this handler
+        if (linkRegex.test(textContent)) {
+          // reset the regex since we don't want the above test method to affect other places where the regex is used
+          linkRegex.lastIndex = 0;
+          return false;
+        }
+
+        if (shiftKey) {
+          return false;
+        }
+
+        if (empty) {
+          if (textContent) {
+            const html = linkifyHtml(textContent);
+            options.editor.commands.insertContent(html);
+            return true;
+          }
+          return false;
+        }
+
+        if (!options.linkOnPaste) {
+          return false;
+        }
```

Comment:
This should be the first condition.

Reply:
That's is what I initially did and since `linkOnPaste` is always `true`, so the check would always run the handler. 

The reason I opted to add this check here is that if we ever decide to disable this option or make it user configurable, it wouldn't affect the earlier logic in the handler of pasting links.

Please let me know if we should move the condition to the top regardless

Reply:
I think moving it to top makes the most sense even if we don't use it.